### PR TITLE
Backtester: USD tracking

### PR DESCRIPTION
**Pull request by gloriousCode**
USA USA

# PR Description

This PR adds USD tracking data to results, establishing a consistent comparison benchmark across exchanges, assets and backtesting results. You can read about the flag [here](https://github.com/gloriousCode/gocryptotrader/tree/totally-total-tracking/backtester/config#strategy-settings) and how tracking currencies are used [here](https://github.com/gloriousCode/gocryptotrader/tree/totally-total-tracking/backtester/funding/trackingcurrencies)
It removes the external API dependency too from the previous PR too.
The feature is enabled by default and can be disabled in strategy configs. It is not supported in live trading data type.

It allows for ratios and performance metrics to be calculated against the combined USD values as an additional way to measure performance:
image

This also comes with the following cool changes:
- Removed Backtesting reliance on existing GoCryptoTrader config file
  - In order to ensure that all pairs are available, the Backtester will run `UpdateTradablePairs` for each exchange as it cannot rely on a config
- New chart provider: Highcharts
- New graphs in report HTML
- Ratio calculation on overall USD results, not just a pair's results
- The return of portfolio interface for strategy analysis. Added two new helper functions to assist:
  - `GetLatestOrderSnapshotForEvent` to retrieve orders for an event's exchange, asset and currency pair
  - `GetLatestOrderSnapshots` to return all orders from the latest snapshots

USD Totals charts:
image
Holdings over time chart:
image
New chart provider can add analysis (in this case RSI)
image

### Things to consider
- The way USD tracking is performed is not 100% accurate as it treats the list of USD currencies [listed here](https://github.com/gloriousCode/gocryptotrader/tree/totally-total-tracking/backtester/funding/trackingcurrencies#what-if-the-exchange-does-not-support-usd) as 1:1 even though they all have minor variances. This is mostly to avoid retrieving significant amounts of data and calculating exchange rates for every candle in order to create a total.
  - This is only for USD total comparisons, a strategy for USDT/BUSD for minor fluctuations will not be converted to 1:1

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How has this been tested
- [x] Test coverage against all new features
- [x] go test ./... -race
- [x] golangci-lint run

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: engine/order_manager.go [hunk 1]
```diff
@@ -527,7 +527,7 @@ func (m *OrderManager) processSubmittedOrder(newOrder *order.Submit, result orde
 		newOrder.Type,
 		newOrder.Date)
 
-	log.Debugln(log.OrderMgr, msg)
+	log.Debugf(log.OrderMgr, msg)
```
Comment (shazbert) [inline engine/order_manager.go:530]: revert?

File: exchanges/kline/kline.go [hunk 2]
```diff
@@ -327,6 +327,9 @@ func TotalCandlesPerInterval(start, end time.Time, interval Interval) (out float
 // IntervalsPerYear helps determine the number of intervals in a year
 // used in CAGR calculation to know the amount of time of an interval in a year
 func (i *Interval) IntervalsPerYear() float64 {
+	if i.Duration() == 0 {
+		return 0
```
Comment (shazbert) [inline exchanges/kline/kline.go:331]: add coverage

**Review by shazbert — CHANGES_REQUESTED** [review]
Everything looks good and seems to work out! Just tried the default DCA strat only one issue I saw on that:

image

Including something on the strategy movement with the same characters. 

Nice work only a few minor nits! 🚀

File: backtester/backtest/backtest.go [hunk 3]
```diff
@@ -130,9 +145,43 @@ func NewFromConfig(cfg *config.Config, templatePath, output string, bot *engine.
 		}
 	}
 
+	var emm = make(map[string]gctexchange.IBotExchange)
+	for i := range cfg.CurrencySettings {
+		_, ok := emm[cfg.CurrencySettings[i].ExchangeName]
+		if ok {
+			continue
```
[the remark this answers is not in the document]
Comment (gloriousCode) [reply backtester/backtest/backtest.go:152]: This map is used to ensure we only setup and call `UpdateTradablePairs(...)` once and added to the backtester's exchange manager per exchange. So when it is `ok`, we've loaded that exchange already and ensured we have all pairs loaded, there should be no errs hidden by this 

I could change it to two loops where I create a slice of exchange names, then the second loop runs  `UpdateTradablePairs(...)` which would be more clear

File: backtester/eventhandlers/exchange/exchange.go [hunk 4]
```diff
@@ -139,7 +138,7 @@ func (e *Exchange) ExecuteOrder(o order.Event, data data.Handler, bot *engine.En
 		funds.IncreaseAvailable(limitReducedAmount.Mul(adjustedPrice), f.GetDirection())
 	}
 
-	ords, _ := bot.OrderManager.GetOrdersSnapshot("")
+	ords, _ := orderManager.GetOrdersSnapshot("")
```
[the remark this answers is not in the document]
Comment (gloriousCode) [reply backtester/eventhandlers/exchange/exchange.go:141]: No error, just a timestamp, but its unused and I'll remove it anyway

File: backtester/eventhandlers/portfolio/portfolio.go [hunk 5]
```diff
@@ -415,17 +440,49 @@ func (p *Portfolio) SetupCurrencySettingsMap(exch string, a asset.Item, cp curre
...
+	return holdings.Holding{}
+}
+
+// Value returns the total value of the latest holdings
+func (e *Settings) Value() decimal.Decimal {
+	latest := e.GetLatestHoldings()
+	if latest.Timestamp.IsZero() {
+		return decimal.Zero
+	}
+	return latest.TotalValue
```
[the remark this answers is not in the document]
Comment (gloriousCode) [reply backtester/eventhandlers/portfolio/portfolio.go:487]: Yeah I removed the function

File: backtester/backtest/backtest.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (shazbert) [reply backtester/backtest/backtest.go:152]: I am happy to keep it how it is, if that's the case.

**Comment by gloriousCode** [comment]
In that screenshot you posted, its from your browser being unable to render `∞`, so I've updated the output to just say `Infinity%` 😄 Microsoft Edge can render it btw................

**Review by shazbert — CHANGES_REQUESTED** [review]
Chart dates are off, mentioned in DM.

**Comment by gloriousCode** [comment]
Totals and holdings charts are now logarithmic per @shazbert's suggestion. Makes reading them a little easier:
image

image

**Review by shazbert — APPROVED** [review]
Tested ACK! Well done! 🚀 x 100. Everything is still deterministic which is lovely. 😆

File: backtester/funding/trackingcurrencies/trackingcurrencies_test.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+		{
+			description:    "currency doesnt exist in available pairs",
+			initialPair:    currency.NewPair(currency.BTC, currency.LTC),
+			availablePairs: &currency.PairStore{Available: currency.Pairs{currency.NewPair(currency.BTC, currency.DOGE)}},
+			basePair:       currency.Pair{},
+			quotePair:      currency.Pair{},
+			expectedErr:    errCurrencyNotFoundInPairs,
+		},
+	}
+	for i := range tests {
```
Comment (thrasher-) [inline backtester/funding/trackingcurrencies/trackingcurrencies_test.go:130]:
```suggestion
	for i := range tests {
		tt := tests[i]
		t.Run(tt.description, func(t *testing.T) {
			t.Parallel()
			basePair, quotePair, err := findMatchingUSDPairs(tt.initialPair, tt.availablePairs)
			if !errors.Is(err, tt.expectedErr) {
				t.Fatalf("'%v' received '%v' expected '%v'", tt.description, err, tt.expectedErr)
			}
			if basePair != tt.basePair {
				t.Fatalf("'%v' received '%v' expected '%v'", tt.description, basePair, tt.basePair)
			}
			if quotePair != tt.quotePair {
				t.Fatalf("'%v' received '%v' expected '%v'", tt.description, quotePair, tt.quotePair)
			}
		})
	}
```

As individually testing this func does not provide coverage

File: exchanges/kline/kline_test.go [hunk 7]
```diff
@@ -981,3 +985,30 @@ func TestConvertToNewInterval(t *testing.T) {
...
+			},
+			{
+				Time:  tt.Add(time.Hour),
+				Close: 1338,
+			},
+		},
+	}
+	price, err := k.GetClosePriceAtTime(tt)
+	if !errors.Is(err, nil) {
+		t.Errorf("received '%v' expectec '%v'", err, nil)
```
Comment (thrasher-) [inline exchanges/kline/kline_test.go:1005]:
```suggestion
		t.Errorf("received '%v' expected '%v'", err, nil)
```
Seems like there's a few instances of these in this file 😄

File: common/convert/convert.go [hunk 8]
```diff
@@ -81,3 +84,83 @@ func BoolPtr(condition bool) *bool {
...
+		number = -number
+		neg = true
+	}
+	str := fmt.Sprintf("%v", number)
+	return numberToCommaSeparatedString(str, 0, "", thousandsSep, neg)
+}
+
+// FloatToCommaSeparatedString converts a float to a comma separated string at the thousand point
+// eg 1000 becomes 1,000
+func FloatToCommaSeparatedString(number float64, decimals uint, decPoint, thousandsSep string) string {
```
Comment (thrasher-) [inline common/convert/convert.go:102]: This func name is technically incorrect as a user can supply whatever they want for the `thousandsSep` param. I think a better name would be "FloatToHumanFriendlyString" or you can drop the thousandsSep param and make it always a comma as the name suggests

File: backtester/funding/funding.go [hunk 9]
```diff
@@ -3,36 +3,43 @@ package funding
...
 )
 
 var (
 	// ErrFundsNotFound used when funds are requested but the funding is not found in the manager
 	ErrFundsNotFound = errors.New("funding not found")
 	// ErrAlreadyExists used when a matching item or pair is already in the funding manager
-	ErrAlreadyExists              = errors.New("funding already exists")
+	ErrAlreadyExists = errors.New("funding already exists")
+	// ErrUSDTrackingDisabled used when attempting to track USD values when disabled
+	ErrUSDTrackingDisabled        = errors.New("usd tracking disabled")
```
Comment (thrasher-) [inline backtester/funding/funding.go:26]:
```suggestion
	ErrUSDTrackingDisabled        = errors.New("USD tracking disabled")
```

**Review by thrasher- — CHANGES_REQUESTED** [review]
Awesome feature! First code pass. Now onto functionality testing 🚀

**Comment by thrasher-** [comment]
Chart functionality is working great! Tested the default DCA as well as the `top2bottom2` strat

image

I noticed an issue with some of the menu items on the chart for things like `save chart`:

image

**Comment by gloriousCode** [comment]
I've update it to only show working/helpful tools on the GUI:
Holdings charts:
image

Candle charts:
image

Also can now save charts:
image

**Review by shazbert — APPROVED** [review]
Changes tested ACK. 🚀

**Review by thrasher- — APPROVED** [review]
changes, tACK! Great stuff @gloriousCode

**Pull request merged**